sretoolbox/utils/cache.py

```diff
@@ -0,0 +1,167 @@
+"""
+Caching helper decorator
+"""
+
+import functools
+import weakref
+
+_CACHES = weakref.WeakKeyDictionary()
+
+
+def static(key):
```

Comment:
I believe users should not have that much control over the cache... I can see people using the same keys in completly unrelated functions.
I know, you want a link between getters and setters, but asking users to do it is not the way to go, IMO.

Reply:
Anyway, key should be unique per function+arguments.

Reply:
I thought of implementing something like
- defaulting the key to the wrapped function's name
- defaulting the key to a derivate of the wrapped function's name like this:
```python
# `func` is the function we wrap/decorate
def _default_static_key(key, func):
    if key is not None:
        return key
    return re.sub(r'[get|set]_', '', func.__name__)
```

but IMHO we should provide a reasonable escape hatch to override this.
Example: 1 of the 2 links provided in the issue uses the function name `_read_all_v2`

---

> key should be unique per function+arguments

I tried to come up with a way for this but I was not able to create a  function `compute_cache_key(func, *args, **kwargs)` that is valid for all ways to call the decorated function.

given

```python
def foo(bar, baz):
  pass
```

i can call it in these ways

```python
foo("hello", "world") # *args = ("hello", "world"), **kwargs = {}
foo(bar="hello", baz="world") # *args = (), **kwargs = {"bar": "hello", "baz": "world"}
foo("hello", baz="world") # *args = ("hello"), **kwargs = {"baz": "world"})
```

which, when 'just' using a tuple of `(func.__name__, args, kwargs)`.

Do you have any thoughts on this?

Reply:
Why do you care? If `foo("hello", "world")` and `foo(bar="hello", baz="world")` are using different backing caches, it's fine.

That will be yet another reason to be consistent . We can even improve ours docs here, including some remarks on that:
https://gitlab.cee.redhat.com/service/dev-guidelines/-/blob/master/python-style-guide.md#L326-352

But having `foo("hello", "world")` and a subsequent `foo("howdy", "folks")` returning the same value cannot happen.